resmem.c

```diff
@@ -50,37 +52,295 @@ static inline int get_info_infile(char *fname, char *res, void *out)
+		 *for that cgroup only.
+		 */
 	case RES_MEM_FREE:
-		return get_info_infile(MEMINFO_FILE, "MemFree:", out);
+		if (cg) {
+			mmusage = cgmemread(cg, "memory.usage_in_bytes");
+			if (get_info_infile(MEMINFO_FILE, "MemTotal:",
+				&memtotal) == -1)
+				return -1;
+			mmtot = cgmemlimit(cg, "memory.limit_in_bytes");
+			if (memtotal < mmtot) {
+				mmtot = memtotal;
+			}
+
+			*(size_t *)out = mmtot - mmusage;
+			return 0;
+		} else {
+			return get_info_infile(MEMINFO_FILE, "MemFree:", out);
+		}
 
 	case RES_MEM_AVAILABLE:
-		return get_info_infile(MEMINFO_FILE, "MemAvailable:", out);
+		if (cg) {
+			mmusage = cgmemread(cg, "memory.usage_in_bytes");
+			if (get_info_infile(MEMINFO_FILE, "MemTotal:",
+				&memtotal) == -1)
+				return -1;
+
+			mmtot = cgmemlimit(cg, "memory.limit_in_bytes");
+
+			if (memtotal < mmtot) {
+				mmtot = memtotal;
+			}
+			cache = cgmemstat(cg, "cache");
+
+			*(size_t *)out = mmtot - mmusage + cache;
+			return 0;
+		} else {
+			return get_info_infile(MEMINFO_FILE, "MemAvailable:",
+				out);
+		}
 
 	case RES_MEM_TOTAL:
-		return get_info_infile(MEMINFO_FILE, "MemTotal:", out);
+		ret = get_info_infile(MEMINFO_FILE, "MemTotal:", out);
+		if (cg) {
+			mmtot = cgmemlimit(cg, "memory.limit_in_bytes");
+			if (ret != -1 && *(size_t *)out > mmtot) {
+				*(size_t *)out = mmtot;
+			}
+		}
+		return ret;
 
 	case RES_MEM_ACTIVE:
-		return get_info_infile(MEMINFO_FILE, "Active:", out);
+		if (cg) {
+			active_anon = cgmemstat(cg, "\nactive_anon");
+			active_file = cgmemstat(cg, "\nactive_file");
+			*(size_t *)out = active_anon + active_file;
+			return 0;
+		} else {
+			return get_info_infile(MEMINFO_FILE, "Active:", out);
+		}
 
 	case RES_MEM_INACTIVE:
-		return get_info_infile(MEMINFO_FILE, "Inactive:", out);
+		if (cg) {
+			inactive_anon = cgmemstat(cg, "\ninactive_anon");
+			inactive_file = cgmemstat(cg, "\ninactive_file");
+			*(size_t *)out = inactive_anon + inactive_file;
+			return 0;
+		} else {
+			return get_info_infile(MEMINFO_FILE, "Inactive:", out);
+		}
 
 	case RES_MEM_SWAPTOTAL:
-		return get_info_infile(MEMINFO_FILE, "SwapTotal:", out);
+		ret = get_info_infile(MEMINFO_FILE, "SwapTotal:", out);
+		if (cg) {
+			swtot = cgmemlimit(cg,
+				"memory.memsw.limit_in_bytes");
+			if (ret != -1 && swtot > 0 && *(size_t *)out > swtot) {
+				*(size_t *)out = swtot;
+			}
+		}
+		return ret;
 
 	case RES_MEM_SWAPFREE:
+		if (cg) {
+			swusage = cgmemread(cg, "memory.memsw.usage_in_bytes");
+			swtot = cgmemlimit(cg,
+				"memory.memsw.limit_in_bytes");
+
+			if (swusage > 0 &&  swtot > 0) {
+				mmusage = cgmemread(cg,
+					"memory.usage_in_bytes");
+				if (get_info_infile(MEMINFO_FILE,
+					"SwapTotal:", &swaptotal) == -1)
+					return -1;
+
+				if (swtot > swaptotal) {
+					swtot = swaptotal;
+				}
+
+				swusage = swusage - mmusage;
+				if (swusage < swtot)
+					swapfree = swtot - swusage;
+				else
+					swapfree = 0;
+				*(size_t *)out = swapfree;
```

Comment:
Probably want to introduce proper pointer casting macros if you intend to do this more often. Since this is non-standard compliant how you do it. :) See LXC again:
https://github.com/lxc/lxc/blob/41cd8a8d8c3e5fd357a4b2fcdceb7b97d70586a4/src/lxc/macro.h#L381

This is really just as trivial as:
```c
#define PTR_TO_SIZE_T(p) ((size_t)((intptr_t)(p)))
```

Reply:
Will check and update.

Reply:
I checked this. However I am not making any pointer to size_t casting. I am changing type of pointer, void * to size_t *, not sure what is the standard way to do that. Is something like
#define PTR_TO_SIZE_T_PTR(p) ((size_t *)((intptr_t)(p)))
is better ?